File: code/HuaWeiNote/app/src/main/java/com/fan/hwnote/app/model/TodoRepository.kt
package com.fan.hwnote.app.model

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import com.fan.hwnote.app.model.db.NoteDbHelper
import com.fan.hwnote.app.model.entity.RepeatType
import com.fan.hwnote.app.model.entity.Todo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.Calendar

object TodoRepository {

    private lateinit var dbHelper: NoteDbHelper

    fun init(context: Context) {
        dbHelper = NoteDbHelper(context.applicationContext)
    }

    suspend fun insert(todo: Todo): Long = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis()
        val cv = toContentValues(todo).apply {
            put("created_at", if (todo.createdAt > 0) todo.createdAt else now)
            put("updated_at", now)
        }
        dbHelper.writableDatabase.insert("todos", null, cv)
    }

    suspend fun update(todo: Todo) = withContext(Dispatchers.IO) {
        val cv = toContentValues(todo).apply {
            put("updated_at", System.currentTimeMillis())
        }
        dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(todo.id.toString()))
        Unit
    }

    suspend fun getById(id: Long): Todo? = withContext(Dispatchers.IO) {
        val cursor = dbHelper.readableDatabase.query(
            "todos", null, "id = ?", arrayOf(id.toString()), null, null, null,
        )
        cursor.use { c -> if (c.moveToFirst()) cursorToTodo(c) else null }
    }

    suspend fun list(
        folderId: Long? = null,
        includeDeleted: Boolean = false,
        hideCompleted: Boolean = false,
    ): List<Todo> = withContext(Dispatchers.IO) {
        val where = mutableListOf<String>()
        val args = mutableListOf<String>()
        if (!includeDeleted) {
            where += "deleted_at = 0"
        } else {
            where += "deleted_at > 0"
        }
        if (folderId != null) {
            where += "folder_id = ?"
            args += folderId.toString()
        }
        if (hideCompleted) {
            where += "is_completed = 0"
        }
        val selection = where.joinToString(" AND ")
        val out = mutableListOf<Todo>()
        val cursor = dbHelper.readableDatabase.query(
            "todos", null, selection, args.toTypedArray(),
            null, null, "remind_at ASC, created_at DESC",
        )
        cursor.use { c -> while (c.moveToNext()) out += cursorToTodo(c) }
        out
    }

    suspend fun listUncategorized(hideCompleted: Boolean = false): List<Todo> =
        withContext(Dispatchers.IO) {
            val where = mutableListOf("deleted_at = 0", "folder_id IS NULL")
            if (hideCompleted) where += "is_completed = 0"
            val selection = where.joinToString(" AND ")
            val out = mutableListOf<Todo>()
            val cursor = dbHelper.readableDatabase.query(
                "todos", null, selection, null, null, null,
                "remind_at ASC, created_at DESC",
            )
            cursor.use { c -> while (c.moveToNext()) out += cursorToTodo(c) }
            out
        }

    suspend fun count(
        folderId: Long? = null,
        includeDeleted: Boolean = false,
    ): Int = withContext(Dispatchers.IO) {
        val where = mutableListOf<String>()
        val args = mutableListOf<String>()
        if (!includeDeleted) {
            where += "deleted_at = 0"
        } else {
            where += "deleted_at > 0"
        }
        if (folderId != null) {
            where += "folder_id = ?"
            args += folderId.toString()
        }
        val selection = where.joinToString(" AND ")
        val cursor = dbHelper.readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM todos WHERE $selection", args.toTypedArray(),
        )
        cursor.use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    suspend fun countUncategorized(): Int = withContext(Dispatchers.IO) {
        val cursor = dbHelper.readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM todos WHERE deleted_at = 0 AND folder_id IS NULL", null,
        )
        cursor.use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
    }

    suspend fun softDelete(id: Long) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("deleted_at", System.currentTimeMillis()) }
        dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun softDeleteBatch(ids: List<Long>) = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis()
        val db = dbHelper.writableDatabase
        db.beginTransaction()
        try {
            val cv = ContentValues().apply { put("deleted_at", now) }
            for (id in ids) {
                db.update("todos", cv, "id = ?", arrayOf(id.toString()))
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
        Unit
    }

    suspend fun restore(id: Long) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply { put("deleted_at", 0L) }
        dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun deletePermanently(id: Long) = withContext(Dispatchers.IO) {
        dbHelper.writableDatabase.delete("todos", "id = ?", arrayOf(id.toString()))
    }

    suspend fun purgeExpired(
        now: Long = System.currentTimeMillis(),
        ttlMs: Long = 30L * 24 * 60 * 60 * 1000,
    ): Int = withContext(Dispatchers.IO) {
        val cutoff = now - ttlMs
        dbHelper.writableDatabase.delete(
            "todos", "deleted_at > 0 AND deleted_at < ?", arrayOf(cutoff.toString()),
        )
    }

    suspend fun completeTodo(id: Long) = withContext(Dispatchers.IO) {
        val todo = getByIdSync(id) ?: return@withContext
        if (todo.repeatType != RepeatType.NONE && todo.remindAt > 0) {
            val nextRemind = advanceRemindAt(todo.remindAt, todo.repeatType)
            val cv = ContentValues().apply {
                put("remind_at", nextRemind)
                put("updated_at", System.currentTimeMillis())
            }
            dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(id.toString()))
        } else {
            val cv = ContentValues().apply {
                put("is_completed", 1)
                put("updated_at", System.currentTimeMillis())
            }
            dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(id.toString()))
        }
    }

    suspend fun uncompleteTodo(id: Long) = withContext(Dispatchers.IO) {
        val cv = ContentValues().apply {
            put("is_completed", 0)
            put("updated_at", System.currentTimeMillis())
        }
        dbHelper.writableDatabase.update("todos", cv, "id = ?", arrayOf(id.toString()))
        Unit
    }

    suspend fun listPendingAlarms(): List<Todo> = withContext(Dispatchers.IO) {
        val out = mutableListOf<Todo>()
        val cursor = dbHelper.readableDatabase.query(
            "todos", null,
            "remind_at > 0 AND is_completed = 0 AND deleted_at = 0",
            null, null, null, "remind_at ASC",
        )
        cursor.use { c -> while (c.moveToNext()) out += cursorToTodo(c) }
        out
    }

    internal fun advanceRemindAt(remindAt: Long, repeatType: RepeatType): Long {
        val now = System.currentTimeMillis()
        val cal = Calendar.getInstance().apply { timeInMillis = remindAt }
        do {
            when (repeatType) {
                RepeatType.DAILY -> cal.add(Calendar.DAY_OF_MONTH, 1)
                RepeatType.WEEKLY -> cal.add(Calendar.DAY_OF_MONTH, 7)
                RepeatType.MONTHLY -> cal.add(Calendar.MONTH, 1)
                RepeatType.YEARLY -> cal.add(Calendar.YEAR, 1)
                RepeatType.NONE -> return remindAt
            }
        } while (cal.timeInMillis < now)
        return cal.timeInMillis
    }

    private fun getByIdSync(id: Long): Todo? {
        val cursor = dbHelper.readableDatabase.query(
            "todos", null, "id = ?", arrayOf(id.toString()), null, null, null,
        )
        return cursor.use { c -> if (c.moveToFirst()) cursorToTodo(c) else null }
    }

    private fun toContentValues(todo: Todo): ContentValues = ContentValues().apply {
        put("title", todo.title)
        put("memo", todo.memo)
        put("is_completed", if (todo.isCompleted) 1 else 0)
        put("is_important", if (todo.isImportant) 1 else 0)
        put("remind_at", todo.remindAt)
        put("repeat_type", todo.repeatType.value)
        if (todo.folderId == null) putNull("folder_id") else put("folder_id", todo.folderId)
        put("deleted_at", todo.deletedAt)
    }

    private fun cursorToTodo(c: Cursor): Todo {
        val folderIdx = c.getColumnIndexOrThrow("folder_id")
        return Todo(
            id = c.getLong(c.getColumnIndexOrThrow("id")),
            title = c.getString(c.getColumnIndexOrThrow("title")) ?: "",
            memo = c.getString(c.getColumnIndexOrThrow("memo")) ?: "",
            isCompleted = c.getInt(c.getColumnIndexOrThrow("is_completed")) == 1,
            isImportant = c.getInt(c.getColumnIndexOrThrow("is_important")) == 1,
            remindAt = c.getLong(c.getColumnIndexOrThrow("remind_at")),
            repeatType = RepeatType.fromValue(
                c.getInt(c.getColumnIndexOrThrow("repeat_type")),
            ),
            folderId = if (c.isNull(folderIdx)) null else c.getLong(folderIdx),
            deletedAt = c.getLong(c.getColumnIndexOrThrow("deleted_at")),
            createdAt = c.getLong(c.getColumnIndexOrThrow("created_at")),
            updatedAt = c.getLong(c.getColumnIndexOrThrow("updated_at")),
        )
    }
}

File: code/HuaweiNoteFlutter/lib/repositories/todo_repository.dart
import 'package:sqflite/sqflite.dart';
import '../db/database_helper.dart';
import '../models/todo.dart';

class TodoRepository {
  final DatabaseHelper _dbHelper;

  TodoRepository(this._dbHelper);

  Future<int> insert(Todo todo) async {
    final db = await _dbHelper.database;
    final now = DateTime.now().millisecondsSinceEpoch;
    final values = _toMap(todo);
    values['created_at'] = todo.createdAt > 0 ? todo.createdAt : now;
    values['updated_at'] = now;
    return db.insert('todos', values);
  }

  Future<void> update(Todo todo) async {
    final db = await _dbHelper.database;
    final values = _toMap(todo);
    values['updated_at'] = DateTime.now().millisecondsSinceEpoch;
    await db.update('todos', values, where: 'id = ?', whereArgs: [todo.id]);
  }

  Future<Todo?> getById(int id) async {
    final db = await _dbHelper.database;
    final rows = await db.query('todos', where: 'id = ?', whereArgs: [id]);
    if (rows.isEmpty) return null;
    return _rowToTodo(rows.first);
  }

  Future<List<Todo>> list({
    int? folderId,
    bool includeDeleted = false,
    bool hideCompleted = false,
  }) async {
    final db = await _dbHelper.database;
    final where = <String>[];
    final args = <Object>[];
    where.add(includeDeleted ? 'deleted_at > 0' : 'deleted_at = 0');
    if (folderId != null) { where.add('folder_id = ?'); args.add(folderId); }
    if (hideCompleted) { where.add('is_completed = 0'); }
    final rows = await db.query('todos',
      where: where.join(' AND '), whereArgs: args,
      orderBy: 'remind_at ASC, created_at DESC',
    );
    return rows.map(_rowToTodo).toList();
  }

  Future<List<Todo>> listUncategorized({bool hideCompleted = false}) async {
    final db = await _dbHelper.database;
    final where = <String>['deleted_at = 0', 'folder_id IS NULL'];
    if (hideCompleted) where.add('is_completed = 0');
    final rows = await db.query('todos',
      where: where.join(' AND '), orderBy: 'remind_at ASC, created_at DESC',
    );
    return rows.map(_rowToTodo).toList();
  }

  Future<int> count({int? folderId, bool includeDeleted = false}) async {
    final db = await _dbHelper.database;
    final where = <String>[];
    final args = <Object>[];
    where.add(includeDeleted ? 'deleted_at > 0' : 'deleted_at = 0');
    if (folderId != null) { where.add('folder_id = ?'); args.add(folderId); }
    final result = await db.rawQuery(
      'SELECT COUNT(*) FROM todos WHERE ${where.join(' AND ')}', args,
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Future<int> countUncategorized() async {
    final db = await _dbHelper.database;
    final result = await db.rawQuery(
      'SELECT COUNT(*) FROM todos WHERE deleted_at = 0 AND folder_id IS NULL',
    );
    return Sqflite.firstIntValue(result) ?? 0;
  }

  Future<void> softDelete(int id) async {
    final db = await _dbHelper.database;
    await db.update('todos', {'deleted_at': DateTime.now().millisecondsSinceEpoch},
      where: 'id = ?', whereArgs: [id]);
  }

  Future<void> softDeleteBatch(List<int> ids) async {
    final db = await _dbHelper.database;
    final now = DateTime.now().millisecondsSinceEpoch;
    await db.transaction((txn) async {
      for (final id in ids) {
        await txn.update('todos', {'deleted_at': now}, where: 'id = ?', whereArgs: [id]);
      }
    });
  }

  Future<void> restore(int id) async {
    final db = await _dbHelper.database;
    await db.update('todos', {'deleted_at': 0}, where: 'id = ?', whereArgs: [id]);
  }

  Future<void> deletePermanently(int id) async {
    final db = await _dbHelper.database;
    await db.delete('todos', where: 'id = ?', whereArgs: [id]);
  }

  Future<void> completeTodo(int id) async {
    final db = await _dbHelper.database;
    final rows = await db.query('todos', where: 'id = ?', whereArgs: [id]);
    if (rows.isEmpty) return;
    final todo = _rowToTodo(rows.first);
    if (todo.repeatType != RepeatType.none && todo.remindAt > 0) {
      final nextRemind = advanceRemindAt(todo.remindAt, todo.repeatType);
      await db.update('todos', {
        'remind_at': nextRemind,
        'updated_at': DateTime.now().millisecondsSinceEpoch,
      }, where: 'id = ?', whereArgs: [id]);
    } else {
      await db.update('todos', {
        'is_completed': 1,
        'updated_at': DateTime.now().millisecondsSinceEpoch,
      }, where: 'id = ?', whereArgs: [id]);
    }
  }

  Future<void> uncompleteTodo(int id) async {
    final db = await _dbHelper.database;
    await db.update('todos', {
      'is_completed': 0,
      'updated_at': DateTime.now().millisecondsSinceEpoch,
    }, where: 'id = ?', whereArgs: [id]);
  }

  Future<List<Todo>> listPendingAlarms() async {
    final db = await _dbHelper.database;
    final rows = await db.query('todos',
      where: 'remind_at > 0 AND is_completed = 0 AND deleted_at = 0',
      orderBy: 'remind_at ASC',
    );
    return rows.map(_rowToTodo).toList();
  }

  Future<int> purgeExpired({
    int? now,
    int ttlMs = 30 * 24 * 60 * 60 * 1000,
  }) async {
    final db = await _dbHelper.database;
    final cutoff = (now ?? DateTime.now().millisecondsSinceEpoch) - ttlMs;
    return db.delete('todos',
      where: 'deleted_at > 0 AND deleted_at < ?', whereArgs: [cutoff]);
  }

  static int advanceRemindAt(int remindAt, RepeatType repeatType) {
    if (repeatType == RepeatType.none) return remindAt;
    final now = DateTime.now().millisecondsSinceEpoch;
    var dt = DateTime.fromMillisecondsSinceEpoch(remindAt);
    do {
      dt = switch (repeatType) {
        RepeatType.daily => dt.add(const Duration(days: 1)),
        RepeatType.weekly => dt.add(const Duration(days: 7)),
        RepeatType.monthly => _addMonth(dt, 1),
        RepeatType.yearly => _addMonth(dt, 12),
        RepeatType.none => dt,
      };
    } while (dt.millisecondsSinceEpoch <= now);
    return dt.millisecondsSinceEpoch;
  }

  static DateTime _addMonth(DateTime dt, int months) {
    final targetYear = dt.year + (dt.month + months - 1) ~/ 12;
    final targetMonth = (dt.month + months - 1) % 12 + 1;
    final lastDay = DateTime(targetYear, targetMonth + 1, 0).day;
    final day = dt.day > lastDay ? lastDay : dt.day;
    return DateTime(targetYear, targetMonth, day, dt.hour, dt.minute, dt.second);
  }

  Map<String, Object?> _toMap(Todo todo) => {
    'title': todo.title,
    'memo': todo.memo,
    'is_completed': todo.isCompleted ? 1 : 0,
    'is_important': todo.isImportant ? 1 : 0,
    'remind_at': todo.remindAt,
    'repeat_type': todo.repeatType.value,
    'folder_id': todo.folderId,
    'deleted_at': todo.deletedAt,
  };

  Todo _rowToTodo(Map<String, Object?> row) => Todo(
    id: row['id'] as int,
    title: row['title'] as String? ?? '',
    memo: row['memo'] as String? ?? '',
    isCompleted: (row['is_completed'] as int) == 1,
    isImportant: (row['is_important'] as int) == 1,
    remindAt: row['remind_at'] as int? ?? 0,
    repeatType: RepeatType.fromValue(row['repeat_type'] as int? ?? 0),
    folderId: row['folder_id'] as int?,
    deletedAt: row['deleted_at'] as int? ?? 0,
    createdAt: row['created_at'] as int,
    updatedAt: row['updated_at'] as int,
  );
}
